### PR TITLE
Feat/support media queries

### PR DESCRIPTION
**Pull request by kaisermann**
#### What problem is this solving?

Breaking the chains of only being able to use device-related media queries.

#### How to test it?

Go to: https://potato--storecomponents.myvtex.com/

WIP

#### Screenshots or example usage:

image

#### Related to / Depends on

https://vtex-dev.atlassian.net/jira/software/projects/SF/boards/105?selectedIssue=SF-248

### CODE REVIEW

**Review by klzns — APPROVED** [review]
It looks great.

I miss some tests though 💔

File: react/hooks/useMediaQueryList.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    (key) => delete BreakpointMatchers[key]
+  )
+}
+
+export function useMediaQueryList(queries: string[]) {
+  const [matches, setMatches] = useState<Record<string, boolean>>(() => {
+    const initial: Record<string, boolean> = {}
+
+    queries.forEach((query) => {
+      let mql = BreakpointMatchers[query]
```
Comment (victorhmp) [inline react/hooks/useMediaQueryList.ts]: Avoid abbreviations to make this bit of code more readable :)

File: react/hooks/useMediaQueryList.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  })
+
+  useLayoutEffect(() => {
+    if (typeof window?.matchMedia !== 'function' || queries.length === 0) {
+      return
+    }
+
+    const cancels = queries.map((query) => {
+      let mounted = true
+      const mql = BreakpointMatchers[query]
```
Comment (victorhmp) [inline react/hooks/useMediaQueryList.ts]: Same thing here

File: react/tests/useResponsiveValue.test.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+jest.mock('vtex.device-detector', () => {
+  return { useDevice: jest.fn() }
+})
+
+function switchDevice(device: string) {
+  return (useDevice as jest.Mock).mockImplementation(() => ({ device }))
+}
+
+describe('useResponsiveValue', () => {
+  devices.forEach((device) => {
```
Comment (klzns) [inline react/tests/useResponsiveValue.test.tsx]: Maybe use test.each instead?
https://jestjs.io/docs/en/api#testeachtablename-fn-timeout

**Review by klzns — APPROVED** [review]

**Review by klzns — CHANGES_REQUESTED** [review]
Build is failing

Maybe remove tests from going to IO
image

**Review by victorhmp — CHANGES_REQUESTED** [review]

File: react/hooks/useMediaQueryList.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+export function useMediaQueryList(queries: string[]) {
+  const [matches, setMatches] = useState<Record<string, boolean>>(() => {
+    const initial: Record<string, boolean> = {}
+
+    queries.forEach((query) => {
+      let mediaQueryList = BreakpointMatchers[query]
+
+      if (!mediaQueryList) {
+        // ugly side-effect（＞д＜）
```
Comment (lbebber) [inline react/hooks/useMediaQueryList.ts]:
```suggestion
```

no need to be apologetic (explain why it was done as a side effect if needed)

**Review by victorhmp — APPROVED** [review]

File: react/hooks/useMediaQueryList.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+export function useMediaQueryList(queries: string[]) {
+  const [matches, setMatches] = useState<Record<string, boolean>>(() => {
+    const initial: Record<string, boolean> = {}
+
+    queries.forEach((query) => {
+      let mediaQueryList = BreakpointMatchers[query]
+
+      if (!mediaQueryList) {
+        // ugly side-effect to get a up-to-date initial value（＞д＜）
```
Comment (lbebber) [inline react/hooks/useMediaQueryList.ts:19]: no need to be apologetic here, it makes sense

**Review by lbebber — APPROVED** [review]
LGTM (left a couple of suggestions over DM but this looks good as it is)

**Review by klzns — APPROVED** [review]

File: docs/README.md [hunk 6]
```diff
@@ -1,9 +1,12 @@
 # vtex.responsive-values
+
 <!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
+
 [![All Contributors](https://img.shields.io/badge/all_contributors-0-orange.svg?style=flat-square)](#contributors-)
+
 <!-- ALL-CONTRIBUTORS-BADGE:END -->
 
-Utility for using props that accept different values for different devices.
+Utility for using props that accept different values for different devices or media queries.
```
Comment (carolinamenezes) [inline docs/README.md:9]:
```suggestion
In VTEX IO, some component's props, such as the ones from the [Flex-Layout](https://developers.vtex.com/docs/vtex-flex-layout) app, support responsive-values. That means it's possible to assign, to the same prop, different values for each device type or specific media queries.

Props that support responsive values accept either a unique value for all cases or an object containing specific definitions for each of the following: `phone`, `tablet`, `desktop`, `mobile` (for both phone and tablet), and media queries.
```

File: docs/README.md [hunk 7]
```diff
@@ -13,13 +16,14 @@ props: {
 }
 ```
 
-...or different values depending on the device:
+...or different values depending on the device or a media query:
```
Comment (carolinamenezes) [inline docs/README.md:19]:
```suggestion
... or different values depending on the device type or specific media queries:
```

File: docs/README.md [hunk 8]
```diff
@@ -80,9 +82,17 @@ const MyComponent = props => {
 /> {/* returns 1 on mobile devices, 2 on desktop */}
 ```
 
+There are some things to keep in mind when using media queries:
+
+- Media queries take precedence over devices. If one matches, any device-specific value is ignored.
+- Media queries are written following the [standard media query syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#Syntax).
+- The definiton order matters. If more than one media query is matched, the first matched one is used.
```
Comment (carolinamenezes) [inline docs/README.md:89]:
```suggestion
Considering the use of a device, which responsive values are missing, the page style will match the following largest screen size definition.

For example, if a component has only `phone` and `desktop` responsive values defined, tablet devices will receive the value from `desktop`. 

Media queries, in turn, have no fallback behavior. Hence, when using this feature, it's important to keep in mind that:

- Media queries take precedence over devices. Therefore, if a media query matches the current state of a window, all device definitions are ignored.
- The order definition matters. Therefore, if more than one media query matches the current state of a window, the first match is used.
- Media queries are written following the [standard media query syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#Syntax).
```

File: docs/README.md [hunk 9]
```diff
@@ -80,9 +82,17 @@ const MyComponent = props => {
...
+
 ## API
```
Comment (carolinamenezes) [inline docs/README.md:92]:
```suggestion
```

File: docs/README.md [hunk 10]
```diff
@@ -80,9 +82,17 @@ const MyComponent = props => {
...
 
-Props that use this hook accept either a value by itself, or an object with the following options: `phone` and `tablet` separately (or just `mobile` for both), and `desktop`. If there is any missing device, it will fallback to the next largest one. For example, if only the values of `phone` and `desktop` are passed, `tablet` devices will receive the value from `desktop`.
+Props that use this hook accept either a value by itself, or an object with the following options: `phone` and `tablet` separately (or just `mobile` for both), `desktop` and generic media queries.
+
+If there is any missing device, it will fallback to the next largest one. For example, if only the values of `phone` and `desktop` are passed, `tablet` devices will receive the value from `desktop`. Media queries have no fallback behavior.
```
Comment (carolinamenezes) [inline docs/README.md:96]:
```suggestion
```

**Pull request merged**